### PR TITLE
[WIP] logs in web console

### LINKED ISSUE #4247: $rootScope.$apply() in DataService.get() causes looping errors
There is a call to `$rooteScope.$apply()` in DataService.get() that can cause Angular to throw `$digest` errors:

`Error: [$rootScope:inprog] $digest already in progress
http://errors.angularjs.org/1.3.8/$rootScope/inprog?p0=%24digest`

A string of promises in controller code is an example of what can trigger this error.  A slimmed down example of what can potentially cause this:

```javascript

AuthService
  .withUser()
  .then(function(user) {
     return DataService
                  .get('projects', projectName, context)
                  .then(function(project) {
                    angular.extend($scope, {
                      project: project
                    });
                    return project;
                  });
  })
  .then(function(user) {
     return DataService
                  .get('builds', buildName, context)
                  .then(function(build) {
                    angular.extend($scope, {
                      build: build
                    })
                    return build;
                  });
  });
```

The block of code that causes the issue (lines 277-292 of data.js):

```javascript
if (obj) {
  $rootScope.$apply(function(){
    deferred.resolve(obj);
  });
}
else {
  $rootScope.$apply(function(){
    // simulation of API object not found
    deferred.reject({
      data: {},
      status: 404,
      headers: function() { return null; },
      config: {}
    });
  });
} 
```

Eliminating the `$rootScope.$apply()` calls does not initially seem to cause issues, but it may be in order to have QE do a regression test.  The block would now look like:

```javascript
if (obj) {
  deferred.resolve(obj);
}
else {
    // simulation of API object not found
  deferred.reject({
      data: {},
      status: 404,
      headers: function() { return null; },
      config: {}
  });
} 
```

The [docs for root scope](https://docs.angularjs.org/error/$rootScope/inprog?p0=$digest) discourage using `$rootScope.$apply()` in this way, particularly under the heading "Digest Phases".  

I propose eliminating these lines unless there is good reason to keep them.  

@jwforres @spadgett any concerns with this change?

### PR DESCRIPTION
**Pull request by benjaminapetersen**
This is a work in progress, but I want to get it out here just in case I end up being out the next couple of weeks.  

Fixes #4247 

@jwforres @spadgett @sg00dwin feel free to comment, thanks!

### CODE REVIEW

File: assets/app/scripts/services/logs/logLinks.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+      var doc = $document[0];
+      return {
+        makeDownload: function(obj) {
+          var a = doc.createElement('a');
+          a.href = $window.URL.createObjectURL(new Blob([obj], {type: 'text/plain'}));
+          a.download = 'log.txt';
+          doc.body.appendChild(a);
+          a.click();
+          doc.body.removeChild(a);
+        },
```
Comment (benjaminapetersen) [inline assets/app/scripts/services/logs/logLinks.js]: d

File: assets/app/scripts/services/logs/logLinks.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+angular.module('openshiftConsole')
+  .factory('logLinks', [
+    '$anchorScroll',
+    '$document',
+    '$location',
+    '$window',
+    function($anchorScroll, $document, $location, $window) {
+      var doc = $document[0];
+      return {
```
Comment (benjaminapetersen) [inline assets/app/scripts/services/logs/logLinks.js]: d

File: assets/app/scripts/services/logs/logLinks.js (hunk 1 above)
Comment (benjaminapetersen) [inline assets/app/scripts/services/logs/logLinks.js]: Here is the 'download plain text file' function.  [Docs for createObjectUrl](https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL).    

Options:
- we just don't use it :smile: 
- I could check prefixes & use it if it works, fallback to noop:
```javascript
var createObjectURL = (window.URL || window.webkitURL || {}).createObjectURL || _.noop;
```

@jwforres @spadgett fire away!

Comment (jwforres) [inline assets/app/scripts/services/logs/logLinks.js]: After reading the doc, the main thing that concerns me is that it doesn't garbage collect those object URLs until the document is unloaded unless you explicitly do a revokeObjectURL call.  Which I think means its keeping a whole extra copy of the log in memory every time you call createObjectURL.  

If we keep this then I think we would need to limit it to only ever creating one per download link (clicking multiple times shouldn't result in multiple objects), and revoke that object URL when we navigate away from the page it was on.  

The download link shouldn't show up at all on browsers where this isn't supported, although from a cursory glance it seems like all the ones we support have it in some form.

Comment (benjaminapetersen) [inline assets/app/scripts/services/logs/logLinks.js]: Yeah, I'm inclined to `URL.revokeObjectURL(objectURL)` immediately after the `doc.body.removeChild(a);`.  The user will already have the file & we don't have to worry about memory.  I can't see a user clicking the link enough times to cause a performance issue, even if they ask for the same log more than once.  Its probably the cleanest way to ensure nothing lingers.

Comment (jwforres) [inline assets/app/scripts/services/logs/logLinks.js]: Ok I wasn't sure if the url object needed to exist until the user is
finished "downloading" the content. But if revoking at that point works
then I agree its the right place to do it.

File: assets/app/views/logs/_log_formatted.html [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+          ng-click="scrollTo('line_'+($index+1), $event)">
+          <span layout flex main-axis="end">
+          {{:: $index+1 }}
+          </span>
+        </a>
+        <span flex class="log-line-text">
+        {{::logLine.text}}
+        </span>
+      </div>
+    </div>
```
Comment (benjaminapetersen) [inline assets/app/views/logs/_log_formatted.html]: Here is where the rendering and filtering has the potential to get a little ugly. 

The current impl works fine for up to 15,000 lines (that I've tested).  There is a delay in render at that point, but we have a loading spinner that makes this clear.  Once rendered with one way binding `{{::var}}` the page does not seem to have issues. Links to top, bottom & individual line #'s work.

 I've got an experimental branch off of this branch where I have tried a couple different angular methods, then a jQuery raw DOM string method and an _.template() method.  There does not seem to be much of a performance improvement, with enough lines each one become sluggish.  

@jwforres @spadgett @sg00dwin

Comment (benjaminapetersen) [inline assets/app/views/logs/_log_formatted.html]: If we want to do more exploration, I'd lean a towards something (angular native) like [Ionic's Collection Repeater](http://ionicframework.com/docs/api/directive/collectionRepeat/) or [Polymer's iron-list](https://elements.polymer-project.org/elements/iron-list).  There is [ng-infinite-scroll](http://binarymuse.github.io/ngInfiniteScroll/demo_basic.html), but this isn't doing the same kind of view recycling, just adding more and more items at the end as the user scrolls.  [Angular UI scroll](http://angular-ui.github.io/ui-scroll/demo/) still stutters when the list gets long. [Angular UI grid](http://ui-grid.info/) seems overkill.

**Comment by spadgett** [comment]
Might need #4247

**Comment by benjaminapetersen** [comment]
I cherry picked it across for the logs & have kept it as a separate commit whenever I rebase logs.

**Comment by spadgett** [comment]
I think we should add a link to the build log here:

screen shot 2015-09-15 at 1 49 19 pm

**Comment by jwforres** [comment]
YES :+1:

Every time I see that I think it would be nice to immediately go see what
went wrong.

**Comment by benjaminapetersen** [comment]
old PR, closing & updating w/a new PR

**Pull request closed without merging**